src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -737,6 +662,23 @@ DistributedInsertSelectSupported(Query *queryTree, RangeTblEntry *insertRte,
 							 NULL, NULL);
 	}
 
+	if (IsCitusTableType(targetRelationId, NULL_KEY_DISTRIBUTED_TABLE) ||
+		subqueryRteListProperties->hasDistTableWithoutShardKey)
+	{
+		List *distributedRelationIdList = DistributedRelationIdList(subquery);
+		distributedRelationIdList = lappend_oid(distributedRelationIdList,
+												targetRelationId);
+
+		if (!AllDistributedRelationsInListColocated(distributedRelationIdList))
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "distributed INSERT ... SELECT cannot reference a "
+								 "distributed table without a shard key together "
+								 "with non-colocated distributed tables",
+								 NULL, NULL);
+		}
```

Comment:
On 12.0, we should expand this to all distributed table types to avoid mistakenly pushing down insert .. select queries that reference non-colocated distributed tables in general. See example #1 in https://github.com/citusdata/citus/issues/6749.